plugins/module_utils/action_module.py

```diff
@@ -0,0 +1,733 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2012-2013 Michael DeHaan <michael.dehaan@gmail.com>
+# Copyright (c) 2016 Toshio Kuratomi <tkuratomi@ansible.com>
+# Copyright (c) 2019 Ansible Project
+# Copyright (c) 2020 Felix Fontein <felix@fontein.de>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+# Parts taken from ansible.module_utils.basic and ansible.module_utils.common.warnings.
```

Comment:
As I understand it, the code in this file is part of Ansible 2.11 and should be dropped from this collection once we don't support Ansible 2.10 any more, right?

Reply:
Not exactly. The code is already in Ansible 2.9, but unfortunately it is part of `AnsibleModule` and cannot be easily reused without hacks. That's why I copy it here. There are plans to extract more of the argument spec validation; once that happens, I hope we can reduce much of the code from this file. Until then, we probably have to live with copies if we don't want to resort to hacks.